p4-16/spec/P4-16-draft-spec.mdk

```diff
@@ -4671,20 +4670,30 @@ controlBody
     ;
 ~ End P4Grammar
 
-Controls cannot be generic (unlike control type declarations); thus
-when used in the context of a `controlDeclaration` the
-`controlTypeDeclaration` cannot contain any type parameters.  I.e.,
-the following declaration is illegal:
+
+It is illegal to instantiate a ```parser``` within a ```control```
+block.
+For a description of the ```optConstructorParameters```, which
+can be used to build parameterized control blocks, see Section
+[#sec-parametrization].
+
+
+Unlike control type declarations, control declarations may not be
+generic---e.g., the following declaration is illegal:
 
 ~ Begin P4Example
 control C<H>(inout H data) { ... }
 ~ End P4Example
 
-For a description of the ```optConstructorParameters```, which can be
-used to build parameterized control blocks, see Section
-[#sec-parametrization].
+Hence, used in the context of a `controlDeclaration` the production
+rule `controlTypeDeclaration` should not yield type parameters.
 
-We start by describing the core components of a ```control``` block,
+P4 does not support exceptional control-flow within a ```control```
+block. That is, there is no equivalent of the ```verify``` statement
+or the ```reject``` state from parsers. Hence, all error handling must
+be performed explicitly by the programmer.
```

Comment:
Perhaps it does not qualify as an exception to the statements above, but 'exit' can be invoked from 'deep within a call stack' of controls and actions, and cause the control block to exit immediately.  It must be explicitly used by the programmer for it to occur, so it is not hidden from them, hence my reason for saying that it is perhaps not needed to be mentioned here.